### PR TITLE
automate login ping tests

### PR DESCRIPTION
**Pull request by Veracity0**
1) Count ping test attempts when you are logging it.
If you get an acceptable ping test on attempt N, only complete login/timein when you unwind back to attempt 1.
2) When timing in to try for a better connection, via timein command or equivalent, only issue logout request if you are currently logged in.
3) When you log in for the first time, if you are in stealth mode (/q), remove that for the purpose of saving actual user name and reading Preferences, as always.
If you did NOT log in with /q, add that to the name field of the LoginRequest, so that all subsequent timeins are done in stealth mode.
4) When doing a ping test at login, if it you got redirected away from api.php (afterlife.php, for example), do not compare the measured ping time with saved historical ping times, which are only api.php.
5) Add 2 new properties:
- pingLoginCount (defaults to 0) the number of times you are willing to let KoLmafia automatically retry ping tests, looking for a successful goal or threshold average. If 0, no automation; a failed test will pop up the usual yes/no/cancel dialog asking if you want to accept, retry, or bail.
- pingLoginFail - logout or login - the action taken if we cannot get a successfull average ping time after pingLoginCount attempts have been made. login means "accept it and remain logged in". logout means "log out in despair" and try again when you timein.
6) These properties are configurable on the Connection Options panel in both the Login Frame and Preferences/General
7) In preferences, that panel has two more lines:
Latest ping test average time was 35 msec.
(That auto-updates each time you execute a ping test - when attempting to timein, for example.)
Try for a better connection: [Time In]
(Press the button to logout (if needed) and time in, using all the options as configured above.

### CODE REVIEW

**Review by heeheehee-kolmafia — APPROVED** [review]

File: src/net/sourceforge/kolmafia/session/LoginManager.java [hunk 1]
```diff
@@ -93,6 +108,40 @@ public static boolean ping() {
...
+        RequestThread.postRequest(new LogoutRequest());
+        return LoginRequest.relogin();
+      }
+      // We've reached our limit
+      StringBuilder buf = new StringBuilder();
+      buf.append("We've tried ");
+      buf.append(String.valueOf(attempt));
+      buf.append(" times to get a fast connection");
+      if (!error.equals("")) {
+        buf.append("- ");
```
[the remark this answers is not in the document]
Comment (Veracity0) [reply src/net/sourceforge/kolmafia/session/LoginManager.java:126]: It should have a leading space.
Fixed.

**Comment by Veracity0** [comment]
I also disabled and enabled the "Time In" button around executing it.

I'm not sure exactly how multiple threads running timeins with ping tests would work, but I'm sure it can't be good.

**Comment by jaadams5** [comment]
Tangentially, ran with this today and the login script only ran once in spite of multiple retries so that has been addressed.  Thank you.

**Comment by Veracity0** [comment]
> Tangentially, ...

Considering that point 1 was specifically intended to address that, it's not exactly a tangent. 🙃

So, do approve this PR?

**Comment by jaadams5** [comment]
> > Tangentially, ...
> So, do you approve this PR?

Real life got in the way and I wasn't 100% certain "only complete login/timein when you unwind back to attempt 1" addressed the issue when I skimmed things.

But now I will make time to review :-)

**Review by jaadams5 — APPROVED** [review]

**Pull request merged**